nomad/state/state_store.go

```diff
@@ -4759,6 +4842,41 @@ func (s *StateStore) updatePluginWithAlloc(index uint64, alloc *structs.Allocati
 	return nil
 }
 
+// updatePluginWithJobSummary updates the CSI plugins for a job when the
+// job summary is updated by an alloc
+func (s *StateStore) updatePluginWithJobSummary(index uint64, summary *structs.JobSummary, alloc *structs.Allocation,
+	txn *memdb.Txn) error {
+
+	ws := memdb.NewWatchSet()
+	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+	if tg == nil {
+		return nil
+	}
+
+	for _, t := range tg.Tasks {
+		if t.CSIPluginConfig != nil {
+			pluginID := t.CSIPluginConfig.ID
+			plug, err := s.CSIPluginByID(ws, pluginID)
+			if err != nil {
+				return err
+			}
+			if plug == nil {
+				plug = structs.NewCSIPlugin(pluginID, index)
```

Comment:
If understand this and the changes in `upsertNodeCSIPlugins` correctly, we'll now see the plugin registered before the fingerprint. (Which I think is a good UX change, in addition to everything else!) It looks like the scheduler should still block placements until the fingerprint is done, but is that only because we don't associate the allocations with the plugin here?

Reply:
Actually, if we switched to recording the allocations here and just setting their fingerprint dependent properties on node update, the scheduler would still block on `ControllersHealthy > 0`. I don't think going that far would be a UX visible change though.

Reply:
Ok that doesn't seem like something we want to do in this pass. Let's keep what you've got here.